OpenId Implementation

lelemm opened the pull request:
Following the PR #1164 

Instruction on how to test this.
- Clone actual server PR and this PR locally, at the server, in the packages.json, point the @actual/web to link to this PR clone (this is important because the pages will come from this package)
This is how I do:
![image](https://github.com/user-attachments/assets/0fc26e80-e4ba-45a9-8f55-8c9043342352)
It's important to remember to BUILD the desktop-client package before starting the server.

- I suggest to start testing this using a empty server instead of an server already configured.

- Once you have actual server running, you can open the URL and start configuring it using this preview doc https://deploy-preview-447.www.actualbudget.org/docs/advanced/oauth-auth

- I suggest start with [configuring from bootstrap](https://deploy-preview-447.www.actualbudget.org/docs/advanced/oauth-auth#configuration-from-bootstrap). Use passwordless.id for testing, probably the easiest way to test.
 
- You can test the openid provider login now. But to test multiple users, you have to configure [multiuser support](https://deploy-preview-447.www.actualbudget.org/docs/advanced/multi-user)

lelemm commented:
@coderabbitai ignore

lelemm commented:
/update-vrt

lelemm commented:
@coderabbitai review

lelemm commented:
@coderabbitai resume

matt-fidd commented:
You might want to pause it again if you make any changes until you're happy for it to do another review, otherwise it can get a bit busy doing it on every commit

lelemm commented:
@coderabbitai ignore

lelemm commented:
> You might want to pause it again if you make any changes until you're happy for it to do another review, otherwise it can get a bit busy doing it on every commit

yeah, there is a bunch of suggestion already. I'm accepting some to fix what code rabbit suggested (some won't compile)

lelemm commented on packages/loot-core/src/server/admin/app.ts:
```diff
@@ -0,0 +1,241 @@
+  return null;
+});
+
+app.method('owner-created', async function () {
+  const res = await get(getServer().BASE_SERVER + '/admin/owner-created/');
+
+  if (res) {
+    return JSON.parse(res) as boolean;
+  }
+
+  return null;
+});
```
this is ok. this route should be public

lelemm commented on packages/loot-core/src/server/admin/app.ts:
```diff
@@ -0,0 +1,241 @@
+// Expose functions to the client
+export const app = createApp<AdminHandlers>();
+
+app.method('user-delete-all', async function (ids) {
+  const userToken = await asyncStorage.getItem('user-token');
+  if (userToken) {
+    try {
+      const res = await del(getServer().BASE_SERVER + '/admin/users', {
+        token: userToken,
+        ids,
+      });
+
+      if (res) {
+        return res;
+      }
+    } catch (err) {
+      return { error: err.reason };
+    }
+  }
+
+  return { someDeletionsFailed: true };
+});
```
fixed

lelemm commented on packages/desktop-client/src/auth/AuthProvider.tsx:
```diff
@@ -0,0 +1,45 @@
+export const AuthProvider = ({ children }: AuthProviderProps) => {
+  const userData = useSelector((state: State) => state.user.data);
+
+  const hasPermission = (permission?: Permissions) => {
+    if (!permission) {
+      return true;
+    }
+
+    return (
+      (userData?.offline ?? false) ||
+      userData?.permission?.toUpperCase() === permission?.toUpperCase()
+    );
+  };
```
fixed this using the serverURL, if there is no server URL ( locally running ), get permission

lelemm commented on packages/desktop-client/src/components/App.tsx:
```diff
@@ -123,6 +127,42 @@ function AppInner() {
     global.Actual.updateAppMenu(budgetId);
   }, [budgetId]);
 
+  useEffect(() => {
+    if (userData?.tokenExpired) {
+      addNotification({
+        type: 'error',
+        id: 'login-expired',
+        title: 'Login expired',
+        sticky: true,
+        message: 'Login expired, please login again.',
+        button: {
+          title: 'Go to login',
+          action: () => {
+            signOut();
+          },
+        },
+      });
+    }
+  }, [userData, userData?.tokenExpired]);
+
+  useEffect(() => {
+    if (userData?.tokenExpired) {
+      addNotification({
+        type: 'error',
+        id: 'login-expired',
+        title: 'Login expired',
+        sticky: true,
+        message: 'Login expired, please login again.',
+        button: {
+          title: 'Go to login',
+          action: () => {
+            signOut();
+          },
+        },
+      });
+    }
+  }, [userData, userData?.tokenExpired]);
```
removed

lelemm commented:
This PR is a nightmare omg lol. went through more than a half of the code rabbit suggestions.
Monday I will continue.

IF anyone want help me with this, checking for string that need translation would be super (and send me a PR for my branch)

lelemm commented:
@coderabbitai ignore

UndyingSoul commented on packages/desktop-client/e2e/settings.test.js-snapshots/Settings-checks-the-page-visuals-1-chromium-linux.png:
Love the idea with enabling OIDC. I think it's a much needed feature. Even though it's feature-flagged, I don't think we should modify the existing layout until its enabled ;)

lelemm commented:
/update-vrt

MikesGlitch reviewed (commented):
Hey @lelemm, do you know if this works on Electron? I remember checking a month or so ago on another PR and it wasn't working due to the auth redirect happening in an external browser window which Electron can't see.

lelemm commented:
> Hey @lelemm, do you know if this works on Electron? I remember checking a month or so ago on another PR and it wasn't working due to the auth redirect happening in an external browser window which Electron can't see.

honestly I didn't test. I was starting to setup my environment to run the electron to test this but I never really went into it

lelemm commented:
/update-vrt

MikesGlitch commented:
@lelemm Is OpenId configured per budget or per server? We have config settings for openid in the server config file, but we're showing the settings for OpenId in the Budget Settings . 

Is it an opt-in kind of thing per budget or if I start using OpenId for one budget do I need to use it for all?

MikesGlitch reviewed (commented):
I found this bug: 
![test](https://github.com/user-attachments/assets/9fb89a2c-3cf9-4136-8916-2329cd60fdfc)


Can you make it so that if the server is already running you don't try to start it again?

I think the server will need to be force closed when navigating away/unmount from the openid config screen as well.

MikesGlitch commented:
I'm still struggling to get this working with Electron. This is what I'm seeing:

![test2](https://github.com/user-attachments/assets/dbf22a70-7624-42c0-939c-17ea9e75969e)

It's opening an external window at my server address (5006), redirecting with the token there and using it on the web. When I navigate back to Electron it doesn't have it. The createOAuthServer stuff starts - I can see the console log - but doesn't pick up on the token. 

I have this branch running with your changes to the server https://github.com/actualbudget/actual-server/pull/523/files

I haven't looked into how desktop apps normally auth with OpenId but I know there's specific flows for it. Just wondering if we're using them?

lelemm commented:
> @lelemm Is OpenId configured per budget or per server? We have config settings for openid in the server config file, but we're showing the settings for OpenId in the Budget Settings .
> 
> Is it an opt-in kind of thing per budget or if I start using OpenId for one budget do I need to use it for all?

At first I created a completely separated dashboard for openid and user management, but during initial reviews, some maintainers asked me to remove the admin dashboard in favor of current existings pages. So openid enable/disable toggle ended inside the settings page.
It IS server-wide, not budget wide

lelemm commented:
> I found this bug: ![test](https://private-user-images.githubusercontent.com/5285928/395884730-9fb89a2c-3cf9-4136-8916-2329cd60fdfc.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzMTQ3NDQsIm5iZiI6MTczNDMxNDQ0NCwicGF0aCI6Ii81Mjg1OTI4LzM5NTg4NDczMC05ZmI4OWEyYy0zY2Y5LTQxMzYtODkxNi0yMzI5Y2Q2MGZkZmMuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMDIwMDQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWQ5ZTRiMjQzNjMxY2M5NWVhNTZjZjIzZTNiMzQ5MDkwOGNmZWVmZjZjN2IwOTdlMWM4MmNlZGI1NDBjY2VmZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.P_3H9pvveL8AxW1E5ebZ-sdkHpBjAD0hmK7nrtsvYG0) [ ![test](https://private-user-images.githubusercontent.com/5285928/395884730-9fb89a2c-3cf9-4136-8916-2329cd60fdfc.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzMTQ3NDQsIm5iZiI6MTczNDMxNDQ0NCwicGF0aCI6Ii81Mjg1OTI4LzM5NTg4NDczMC05ZmI4OWEyYy0zY2Y5LTQxMzYtODkxNi0yMzI5Y2Q2MGZkZmMuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMDIwMDQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWQ5ZTRiMjQzNjMxY2M5NWVhNTZjZjIzZTNiMzQ5MDkwOGNmZWVmZjZjN2IwOTdlMWM4MmNlZGI1NDBjY2VmZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.P_3H9pvveL8AxW1E5ebZ-sdkHpBjAD0hmK7nrtsvYG0) ](https://private-user-images.githubusercontent.com/5285928/395884730-9fb89a2c-3cf9-4136-8916-2329cd60fdfc.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzMTQ3NDQsIm5iZiI6MTczNDMxNDQ0NCwicGF0aCI6Ii81Mjg1OTI4LzM5NTg4NDczMC05ZmI4OWEyYy0zY2Y5LTQxMzYtODkxNi0yMzI5Y2Q2MGZkZmMuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMDIwMDQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWQ5ZTRiMjQzNjMxY2M5NWVhNTZjZjIzZTNiMzQ5MDkwOGNmZWVmZjZjN2IwOTdlMWM4MmNlZGI1NDBjY2VmZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.P_3H9pvveL8AxW1E5ebZ-sdkHpBjAD0hmK7nrtsvYG0) [ ](https://private-user-images.githubusercontent.com/5285928/395884730-9fb89a2c-3cf9-4136-8916-2329cd60fdfc.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzMTQ3NDQsIm5iZiI6MTczNDMxNDQ0NCwicGF0aCI6Ii81Mjg1OTI4LzM5NTg4NDczMC05ZmI4OWEyYy0zY2Y5LTQxMzYtODkxNi0yMzI5Y2Q2MGZkZmMuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMDIwMDQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWQ5ZTRiMjQzNjMxY2M5NWVhNTZjZjIzZTNiMzQ5MDkwOGNmZWVmZjZjN2IwOTdlMWM4MmNlZGI1NDBjY2VmZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.P_3H9pvveL8AxW1E5ebZ-sdkHpBjAD0hmK7nrtsvYG0)
> 
> Can you make it so that if the server is already running you don't try to start it again?
> 
> I think the server will need to be force closed when navigating away/unmount from the openid config screen as well.

the server should die once you did login once. let me check this

lelemm commented:
> I'm still struggling to get this working with Electron. This is what I'm seeing:
> 
> ![test2](https://private-user-images.githubusercontent.com/5285928/395886341-dbf22a70-7624-42c0-939c-17ea9e75969e.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzMTQ3NDMsIm5iZiI6MTczNDMxNDQ0MywicGF0aCI6Ii81Mjg1OTI4LzM5NTg4NjM0MS1kYmYyMmE3MC03NjI0LTQyYzAtOTM5Yy0xN2VhOWU3NTk2OWUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMDIwMDQzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTEwZjdiMTQzMzUxZWNkODUzMWY4ZWRlMDAwYzIzN2M2YTg2YmUzYWM0ZDBmODFhNGZmOWNkNzEwZjhmMDMzOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Iwu1CPr-_oVaLCKsbz7qKmSoFAveC_ukT3ePdTdgZ3E) [ ![test2](https://private-user-images.githubusercontent.com/5285928/395886341-dbf22a70-7624-42c0-939c-17ea9e75969e.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzMTQ3NDMsIm5iZiI6MTczNDMxNDQ0MywicGF0aCI6Ii81Mjg1OTI4LzM5NTg4NjM0MS1kYmYyMmE3MC03NjI0LTQyYzAtOTM5Yy0xN2VhOWU3NTk2OWUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMDIwMDQzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTEwZjdiMTQzMzUxZWNkODUzMWY4ZWRlMDAwYzIzN2M2YTg2YmUzYWM0ZDBmODFhNGZmOWNkNzEwZjhmMDMzOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Iwu1CPr-_oVaLCKsbz7qKmSoFAveC_ukT3ePdTdgZ3E) ](https://private-user-images.githubusercontent.com/5285928/395886341-dbf22a70-7624-42c0-939c-17ea9e75969e.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzMTQ3NDMsIm5iZiI6MTczNDMxNDQ0MywicGF0aCI6Ii81Mjg1OTI4LzM5NTg4NjM0MS1kYmYyMmE3MC03NjI0LTQyYzAtOTM5Yy0xN2VhOWU3NTk2OWUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMDIwMDQzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTEwZjdiMTQzMzUxZWNkODUzMWY4ZWRlMDAwYzIzN2M2YTg2YmUzYWM0ZDBmODFhNGZmOWNkNzEwZjhmMDMzOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Iwu1CPr-_oVaLCKsbz7qKmSoFAveC_ukT3ePdTdgZ3E) [ ](https://private-user-images.githubusercontent.com/5285928/395886341-dbf22a70-7624-42c0-939c-17ea9e75969e.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzMTQ3NDMsIm5iZiI6MTczNDMxNDQ0MywicGF0aCI6Ii81Mjg1OTI4LzM5NTg4NjM0MS1kYmYyMmE3MC03NjI0LTQyYzAtOTM5Yy0xN2VhOWU3NTk2OWUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMDIwMDQzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTEwZjdiMTQzMzUxZWNkODUzMWY4ZWRlMDAwYzIzN2M2YTg2YmUzYWM0ZDBmODFhNGZmOWNkNzEwZjhmMDMzOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Iwu1CPr-_oVaLCKsbz7qKmSoFAveC_ukT3ePdTdgZ3E)
> 
> It's opening an external window at my server address (5006), redirecting with the token there and using it on the web. When I navigate back to Electron it doesn't have it. The createOAuthServer stuff starts - I can see the console log - but doesn't pick up on the token.
> 
> I have this branch running with your changes to the server https://github.com/actualbudget/actual-server/pull/523/files
> 
> I haven't looked into how desktop apps normally auth with OpenId but I know there's specific flows for it. Just wondering if we're using them? This seems like a nice example of OpenId with Electron: https://auth0.com/blog/securing-electron-applications-with-openid-connect-and-oauth-2/

This happens because the server must have the updated frontend inside it.
![image](https://github.com/user-attachments/assets/f4e6187b-8243-4704-830a-db983d99cb55)

![image](https://github.com/user-attachments/assets/fa761254-d164-4927-b282-0ab927c6cbea)

Remember to build the desktop-client package to generate the dist files.
If you try to use the 24.12.0 desktop-client, it will keep you in a loop

lelemm commented:
made a small change to prevent trying to create multiple servers @MikesGlitch

MikesGlitch commented:
> Remember to build the desktop-client package to generate the dist files. If you try to use the 24.12.0 desktop-client, it will keep you in a loop

Forgot about that - thanks for the reminder. Looks good. It's now working: 

![image](https://github.com/user-attachments/assets/ef6dfab2-6f28-449b-b473-4e20b2e8b84f)

One weird thing I noticed:
![image](https://github.com/user-attachments/assets/98524881-1139-4222-8659-06688bd1b614)

I'm not sure why some of these files have owner: unknown and some don't have any value at all. Instead of _unknown_ maybe we should change it to _Unassigned_ (or something similar) so it looks like less of an error/wrong value? Unknown suggests to me that it's set but it's an unknown value.

MikesGlitch commented:
I found this issue, it happens when you open a budget, then close it:

![test23](https://github.com/user-attachments/assets/61e3aac6-c1a3-455d-a167-08a725dd1855)

lelemm commented:
I changed the name for file owners and also changed solved the bug from privacy filter @MikesGlitch 
Also, I just noticed that the container of the file turned into a button
![image](https://github.com/user-attachments/assets/d59b71f9-dc88-4d7d-89d3-b236d0c16ae9)


Now react is not happy:
![image](https://github.com/user-attachments/assets/9160c4fd-7b9a-49c7-97fe-23802e136723)

lelemm commented:
> Now react is not happy: ![image](https://private-user-images.githubusercontent.com/15043428/396189447-9160c4fd-7b9a-49c7-97fe-23802e136723.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzNjg2ODgsIm5iZiI6MTczNDM2ODM4OCwicGF0aCI6Ii8xNTA0MzQyOC8zOTYxODk0NDctOTE2MGM0ZmQtN2I5YS00OWM3LTk3ZmUtMjM4MDJlMTM2NzIzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE2VDE2NTk0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE4ZWZlMzY0Mjc0NjMwMDUzOWM4MTQwZTM3MTllODBhYzFjZTg2YTJmZTRjYjFiYzJkZTRmMTY3ZmI4OTFiNmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uIKKYMjT6s7KEisjSRP5IyfDPpvXVhTezzNYqYCP7Ec)

oh, that's o master too:
https://github.com/actualbudget/actual/blob/8446356cc62e1311b08627b22cf16a4e46c414c6/packages/desktop-client/src/components/manager/BudgetList.tsx#L109
https://github.com/actualbudget/actual/blob/8446356cc62e1311b08627b22cf16a4e46c414c6/packages/desktop-client/src/components/manager/BudgetList.tsx#L213

MikesGlitch commented:
> > Now react is not happy: 

That's been on master for a while, it's an issue but nothing to do wth this PR

MikesGlitch commented on packages/desktop-electron/index.ts:
```diff
@@ -54,6 +55,37 @@ if (!isDev || !process.env.ACTUAL_DATA_DIR) {
+
+      const code = query.get('token');
+      if (code && clientWin) {
+        clientWin.loadURL(`http://localhost:3001/openid-cb?token=${code}`);
```
I've just confirmed this: 

![image](https://github.com/user-attachments/assets/ba6eeeb3-9308-4d2a-ab1a-dcfbda7a8836)

localhost:3001 isn't running when running in prod mode. This will need to account for the running serverurl. In prod it should be app://actual I think

lelemm replied:
ops, I messed this up. I will change it for the app url

lelemm replied:
I'm trying to build electron to test it, but my environment is not ready for this (idk what's missing).

but I guess this would be the change, right?

```typescript
        if (isDev) {
          clientWin.loadURL(`http://localhost:3001/openid-cb?token=${code}`);
        } else {
          clientWin.loadURL(`app://actual/openid-cb?token=${code}`);
        }
```

MikesGlitch replied:
Probably. You can test by downloading the artifact on this PR.

MikesGlitch replied:
That solved it 🥳

lelemm replied:
I was downloading the artifact to test too just in case

lelemm replied:
working on windows build too 🎊

lelemm commented on packages/desktop-client/src/components/manager/subscribe/OpenIdForm.tsx:
```diff
@@ -0,0 +1,448 @@
+    {
+      label: 'Microsoft Entra',
+      value: 'microsoft',
+      issuer: 'https://login.microsoftonline.com/common/v2.0/',
```
tested both. both worked. I will update for the new one

MikesGlitch replied:
That's weird, when I used the one here it gave me the warning about it being deprecated. Maybe it's to do with when the directory was created or something. Mine is fairly new.

lelemm replied:
mine is from when I built the list

MikesGlitch commented:
I setup the OpenId stuff, then disabled it in the settings page. Now when I go back to the budget list page I see this:

![image](https://github.com/user-attachments/assets/b5fedad6-28c8-4654-81f0-33de020af283)

I'm thinking we should hide the "Logged in as: " part If the logged in user isn't known?

lelemm commented:
> I setup the OpenId stuff, then disabled it in the settings page. Now when I go back to the budget list page I see this:
> 
> ![image](https://private-user-images.githubusercontent.com/5285928/396260796-b5fedad6-28c8-4654-81f0-33de020af283.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzODE4OTAsIm5iZiI6MTczNDM4MTU5MCwicGF0aCI6Ii81Mjg1OTI4LzM5NjI2MDc5Ni1iNWZlZGFkNi0yOGM4LTQ2NTQtODFmMC0zM2RlMDIwYWYyODMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjAzOTUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2EzZjJjYTE2YmIxZWJjM2VmYzgwMzA0NTdkMjgzYjU1ZDUzNzNiYTZlNDFjNTVkYTQwNzc3YzFjYzc4NzBiNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.wgHnVzJrsyLPdGOQz36ES27Mv4MN6mUywiV_UPpsTeI)
> 
> I'm thinking we should hide the "Logged in as: " part If the logged in user isn't known?

Didn't you get redirected to login again? When disabling openid, it should clear your token, redirect to login and when login in again, it should not show "logged as"

Were you using electron client?

lelemm commented:
> How do I get rid of the Unassigned and make it say my username?
> 
> I was logged in when I did this:
> 
> ![unassigned](https://private-user-images.githubusercontent.com/5285928/396262797-fd6fa766-ba7c-4886-8c34-4d87773cca91.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzODIxOTMsIm5iZiI6MTczNDM4MTg5MywicGF0aCI6Ii81Mjg1OTI4LzM5NjI2Mjc5Ny1mZDZmYTc2Ni1iYTdjLTQ4ODYtOGMzNC00ZDg3NzczY2NhOTEuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjA0NDUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmI5MGMzZDgxYzdjNzJkMjc3OWYwY2IxZjJmZTFkYWVmNmJiOTU1YmFhN2VlOGM2MTE4YmE5NmYxYjRhN2M0NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.fEh6cAJi7uE58SapBA1H84qQL0GqZ-4XtsbomGPXprA) [ ![unassigned](https://private-user-images.githubusercontent.com/5285928/396262797-fd6fa766-ba7c-4886-8c34-4d87773cca91.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzODIxOTMsIm5iZiI6MTczNDM4MTg5MywicGF0aCI6Ii81Mjg1OTI4LzM5NjI2Mjc5Ny1mZDZmYTc2Ni1iYTdjLTQ4ODYtOGMzNC00ZDg3NzczY2NhOTEuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjA0NDUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmI5MGMzZDgxYzdjNzJkMjc3OWYwY2IxZjJmZTFkYWVmNmJiOTU1YmFhN2VlOGM2MTE4YmE5NmYxYjRhN2M0NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.fEh6cAJi7uE58SapBA1H84qQL0GqZ-4XtsbomGPXprA) ](https://private-user-images.githubusercontent.com/5285928/396262797-fd6fa766-ba7c-4886-8c34-4d87773cca91.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzODIxOTMsIm5iZiI6MTczNDM4MTg5MywicGF0aCI6Ii81Mjg1OTI4LzM5NjI2Mjc5Ny1mZDZmYTc2Ni1iYTdjLTQ4ODYtOGMzNC00ZDg3NzczY2NhOTEuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjA0NDUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmI5MGMzZDgxYzdjNzJkMjc3OWYwY2IxZjJmZTFkYWVmNmJiOTU1YmFhN2VlOGM2MTE4YmE5NmYxYjRhN2M0NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.fEh6cAJi7uE58SapBA1H84qQL0GqZ-4XtsbomGPXprA) [ ](https://private-user-images.githubusercontent.com/5285928/396262797-fd6fa766-ba7c-4886-8c34-4d87773cca91.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzODIxOTMsIm5iZiI6MTczNDM4MTg5MywicGF0aCI6Ii81Mjg1OTI4LzM5NjI2Mjc5Ny1mZDZmYTc2Ni1iYTdjLTQ4ODYtOGMzNC00ZDg3NzczY2NhOTEuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjA0NDUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmI5MGMzZDgxYzdjNzJkMjc3OWYwY2IxZjJmZTFkYWVmNmJiOTU1YmFhN2VlOGM2MTE4YmE5NmYxYjRhN2M0NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.fEh6cAJi7uE58SapBA1H84qQL0GqZ-4XtsbomGPXprA)
> 
> I thought being logged in and being the first person to view the budget would assign me but no luck.

Click on this menu:
![image](https://github.com/user-attachments/assets/231e24fc-d373-41ac-9c56-eb7da8b80277)

Then click on the 3 dots:
![image](https://github.com/user-attachments/assets/ab82c9c1-96db-43f7-9479-f1d8a58c3ca7)

Then click on the padlock, select your user then click on transfer ownership

MikesGlitch commented:
Found a weird one. Now when I try to disable OpenID I get an internal server error. 

![weird](https://github.com/user-attachments/assets/1bb3a95b-4f44-4047-824a-589fc6f1052a)

lelemm commented:
> padlock

Weird indeed. if you can debug the server to give more information would be lovely

MikesGlitch commented:
> Weird indeed. if you can debug the server to give more information would be lovely

It's ```invalid-password``` error. We should handle that a bit more gracefully - internal server error looked like something broke. 

Right now the server returns a 500 instead of a 401. We should hook inito a 401 on the client and say the user's entered the wrong password IMO.

MikesGlitch commented:
There was some new functionality added recently for _Duplicate budget_.  I noticed it's not duplicating/showing the owner field.

![duplicate](https://github.com/user-attachments/assets/29b98a8e-3281-4417-abc0-47d7bf31a124)

When I set the owner on a duplicated budget, it also doesn't show on the list view.

lelemm commented:
> There was some new functionality added recently for _Duplicate budget_. I noticed it's not duplicating/showing the owner field.
> 
> ![duplicate](https://private-user-images.githubusercontent.com/5285928/396269467-29b98a8e-3281-4417-abc0-47d7bf31a124.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzOTAwNTgsIm5iZiI6MTczNDM4OTc1OCwicGF0aCI6Ii81Mjg1OTI4LzM5NjI2OTQ2Ny0yOWI5OGE4ZS0zMjgxLTQ0MTctYWJjMC00N2Q3YmYzMWExMjQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjI1NTU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjE4M2Q2MTFhZWRjMzg3YzAzZWJkZGY5NmM1YjY1MWRjZjU3YTZiOWYwNTY0NjQwNWZiZTY0ZjI2ZTZjNWM5YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.2a1YDk8WYx2J8_Oxs4QsqT9roAClS9ESQ7Bds5eqiuI) [ ![duplicate](https://private-user-images.githubusercontent.com/5285928/396269467-29b98a8e-3281-4417-abc0-47d7bf31a124.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzOTAwNTgsIm5iZiI6MTczNDM4OTc1OCwicGF0aCI6Ii81Mjg1OTI4LzM5NjI2OTQ2Ny0yOWI5OGE4ZS0zMjgxLTQ0MTctYWJjMC00N2Q3YmYzMWExMjQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjI1NTU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjE4M2Q2MTFhZWRjMzg3YzAzZWJkZGY5NmM1YjY1MWRjZjU3YTZiOWYwNTY0NjQwNWZiZTY0ZjI2ZTZjNWM5YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.2a1YDk8WYx2J8_Oxs4QsqT9roAClS9ESQ7Bds5eqiuI) ](https://private-user-images.githubusercontent.com/5285928/396269467-29b98a8e-3281-4417-abc0-47d7bf31a124.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzOTAwNTgsIm5iZiI6MTczNDM4OTc1OCwicGF0aCI6Ii81Mjg1OTI4LzM5NjI2OTQ2Ny0yOWI5OGE4ZS0zMjgxLTQ0MTctYWJjMC00N2Q3YmYzMWExMjQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjI1NTU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjE4M2Q2MTFhZWRjMzg3YzAzZWJkZGY5NmM1YjY1MWRjZjU3YTZiOWYwNTY0NjQwNWZiZTY0ZjI2ZTZjNWM5YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.2a1YDk8WYx2J8_Oxs4QsqT9roAClS9ESQ7Bds5eqiuI) [ ](https://private-user-images.githubusercontent.com/5285928/396269467-29b98a8e-3281-4417-abc0-47d7bf31a124.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzOTAwNTgsIm5iZiI6MTczNDM4OTc1OCwicGF0aCI6Ii81Mjg1OTI4LzM5NjI2OTQ2Ny0yOWI5OGE4ZS0zMjgxLTQ0MTctYWJjMC00N2Q3YmYzMWExMjQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjI1NTU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjE4M2Q2MTFhZWRjMzg3YzAzZWJkZGY5NmM1YjY1MWRjZjU3YTZiOWYwNTY0NjQwNWZiZTY0ZjI2ZTZjNWM5YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.2a1YDk8WYx2J8_Oxs4QsqT9roAClS9ESQ7Bds5eqiuI)
> 
> When I set the owner on a duplicated budget, it also doesn't show on the list view.

When you duplicate, it creates a new file in the server, the new file is always set to the logged user. So, in this case, you are the owner of this copy.
When you are the owner of a copy, it hides the field "Owner"
![image](https://github.com/user-attachments/assets/12563fce-7a0f-40f9-8e3a-1a9ebaf6776f)

A maintainer (I don't remember who, and I couldn't find in discord the message) asked me for this, when you are the owner, hide the "Owner: ...". It used to show "Owner: You" before this change

lelemm commented:
> There was some new functionality added recently for _Duplicate budget_. I noticed it's not duplicating/showing the owner field.
> 
> ![duplicate](https://private-user-images.githubusercontent.com/5285928/396269467-29b98a8e-3281-4417-abc0-47d7bf31a124.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzODM3OTcsIm5iZiI6MTczNDM4MzQ5NywicGF0aCI6Ii81Mjg1OTI4LzM5NjI2OTQ2Ny0yOWI5OGE4ZS0zMjgxLTQ0MTctYWJjMC00N2Q3YmYzMWExMjQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjExMTM3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjAxMzQ4NzJiMGU4NDkyMzYyYWMzNDk5MjZmZjY0NDEzYTljODRhYjkzOGI4ZWExNTAxOGRkYmIyNzM5YTc0YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.e1UI9HRba-5JOqyLTEKKppGTv_hf22I4jtbDmSNshzw) [ ![duplicate](https://private-user-images.githubusercontent.com/5285928/396269467-29b98a8e-3281-4417-abc0-47d7bf31a124.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzODM3OTcsIm5iZiI6MTczNDM4MzQ5NywicGF0aCI6Ii81Mjg1OTI4LzM5NjI2OTQ2Ny0yOWI5OGE4ZS0zMjgxLTQ0MTctYWJjMC00N2Q3YmYzMWExMjQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjExMTM3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjAxMzQ4NzJiMGU4NDkyMzYyYWMzNDk5MjZmZjY0NDEzYTljODRhYjkzOGI4ZWExNTAxOGRkYmIyNzM5YTc0YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.e1UI9HRba-5JOqyLTEKKppGTv_hf22I4jtbDmSNshzw) ](https://private-user-images.githubusercontent.com/5285928/396269467-29b98a8e-3281-4417-abc0-47d7bf31a124.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzODM3OTcsIm5iZiI6MTczNDM4MzQ5NywicGF0aCI6Ii81Mjg1OTI4LzM5NjI2OTQ2Ny0yOWI5OGE4ZS0zMjgxLTQ0MTctYWJjMC00N2Q3YmYzMWExMjQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjExMTM3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjAxMzQ4NzJiMGU4NDkyMzYyYWMzNDk5MjZmZjY0NDEzYTljODRhYjkzOGI4ZWExNTAxOGRkYmIyNzM5YTc0YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.e1UI9HRba-5JOqyLTEKKppGTv_hf22I4jtbDmSNshzw) [ ](https://private-user-images.githubusercontent.com/5285928/396269467-29b98a8e-3281-4417-abc0-47d7bf31a124.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzODM3OTcsIm5iZiI6MTczNDM4MzQ5NywicGF0aCI6Ii81Mjg1OTI4LzM5NjI2OTQ2Ny0yOWI5OGE4ZS0zMjgxLTQ0MTctYWJjMC00N2Q3YmYzMWExMjQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjExMTM3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjAxMzQ4NzJiMGU4NDkyMzYyYWMzNDk5MjZmZjY0NDEzYTljODRhYjkzOGI4ZWExNTAxOGRkYmIyNzM5YTc0YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.e1UI9HRba-5JOqyLTEKKppGTv_hf22I4jtbDmSNshzw)
> 
> When I set the owner on a duplicated budget, it also doesn't show on the list view.



> > Weird indeed. if you can debug the server to give more information would be lovely
> 
> It's `invalid-password` error. We should handle that a bit more gracefully - internal server error looked like something broke.
> 
> Right now the server returns a 500 instead of a 401. We should hook inito a 401 on the client and say the user's entered the wrong password IMO.

Created an actual server PR for this case:
https://github.com/actualbudget/actual-server/pull/526

lelemm commented:
> Found a weird one. Now when I try to disable OpenID I get an internal server error.
> 
> ![weird](https://private-user-images.githubusercontent.com/5285928/396264250-1bb3a95b-4f44-4047-824a-589fc6f1052a.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzOTM2MDcsIm5iZiI6MTczNDM5MzMwNywicGF0aCI6Ii81Mjg1OTI4LzM5NjI2NDI1MC0xYmIzYTk1Yi00ZjQ0LTQwNDctODI0YS01ODlmYzZmMTA1MmEuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjM1NTA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGIwMmUzMTAxMTg0MTQ2YWMwODEwYjg5MDA1YWMzYjljNzZiMTQ5NTkwOTk2OTA2ZDRjZWNjYjI5NDM4Y2E0MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ddBsB_iAloilfF6ml8OhZ1Z8wHnTJpU974GPo_iHjUY) [ ![weird](https://private-user-images.githubusercontent.com/5285928/396264250-1bb3a95b-4f44-4047-824a-589fc6f1052a.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzOTM2MDcsIm5iZiI6MTczNDM5MzMwNywicGF0aCI6Ii81Mjg1OTI4LzM5NjI2NDI1MC0xYmIzYTk1Yi00ZjQ0LTQwNDctODI0YS01ODlmYzZmMTA1MmEuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjM1NTA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGIwMmUzMTAxMTg0MTQ2YWMwODEwYjg5MDA1YWMzYjljNzZiMTQ5NTkwOTk2OTA2ZDRjZWNjYjI5NDM4Y2E0MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ddBsB_iAloilfF6ml8OhZ1Z8wHnTJpU974GPo_iHjUY) ](https://private-user-images.githubusercontent.com/5285928/396264250-1bb3a95b-4f44-4047-824a-589fc6f1052a.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzOTM2MDcsIm5iZiI6MTczNDM5MzMwNywicGF0aCI6Ii81Mjg1OTI4LzM5NjI2NDI1MC0xYmIzYTk1Yi00ZjQ0LTQwNDctODI0YS01ODlmYzZmMTA1MmEuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjM1NTA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGIwMmUzMTAxMTg0MTQ2YWMwODEwYjg5MDA1YWMzYjljNzZiMTQ5NTkwOTk2OTA2ZDRjZWNjYjI5NDM4Y2E0MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ddBsB_iAloilfF6ml8OhZ1Z8wHnTJpU974GPo_iHjUY) [ ](https://private-user-images.githubusercontent.com/5285928/396264250-1bb3a95b-4f44-4047-824a-589fc6f1052a.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzOTM2MDcsIm5iZiI6MTczNDM5MzMwNywicGF0aCI6Ii81Mjg1OTI4LzM5NjI2NDI1MC0xYmIzYTk1Yi00ZjQ0LTQwNDctODI0YS01ODlmYzZmMTA1MmEuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjM1NTA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGIwMmUzMTAxMTg0MTQ2YWMwODEwYjg5MDA1YWMzYjljNzZiMTQ5NTkwOTk2OTA2ZDRjZWNjYjI5NDM4Y2E0MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ddBsB_iAloilfF6ml8OhZ1Z8wHnTJpU974GPo_iHjUY)

fixed

lelemm commented:
I'm reviewing the docs I generated months ago and I noticed that during code review of the server I removed a important feature from the server by mistake. To use enable openid from the modal in the settings, the openid provider must have a discovery endpoint. Some providers do not have it (like github), you have to setup each endpoint manually. The way I did, this is done using the config.json from the server.


```
  openId?: {
    issuer:
      | string
      | {
          name: string;
          authorization_endpoint: string;
          token_endpoint: string;
          userinfo_endpoint: string;
        };
    client_id: string;
    client_secret: string;
    server_hostname: string;
  };
```

But this feature is useless as it is atm because I removed by mistake the code that would check for this config at server start.

I have to put this feature back to the server code to enable this other type of configuration.
But instead of checking every server start, I'm want to do a separated command in the packages.json much like the reset-password:
![image](https://github.com/user-attachments/assets/62e765a8-bba0-4d9d-a2d0-60d97485a448)

so the user can "npm run enable-openid" from the server, then the server uses the config.json configuration and enable it.

WDYT @matt-fidd / @MikesGlitch / @youngcw ?

MikesGlitch commented:
> so the user can "npm run enable-openid" from the server, then the server uses the config.json configuration and enable it.

I'm a bit slow, what would the script do? Does it walk you through setting up the config.json with the openid info? 

If so, I'm for this approach - it might make it simpler for some people.

lelemm commented:
> > so the user can "npm run enable-openid" from the server, then the server uses the config.json configuration and enable it.
> 
> What would the script do? Does it walk you through setting up the config.json with the openid info?
> 
> If so, I'm for this approach - it might make it simpler.

This script would trigger what the UI triggers on the server, but instead of passing each parameter, it would look for the values in the environment or config.json
It would not walk the user through, but instead validate and enable based on the environment/config

MikesGlitch commented:
> This script would trigger what the UI triggers on the server, but instead of passing each parameter, it would look for the values in the environment or config.json It would not walk the user through, but instead validate and enable based on the environment/config

What's the downside of loading the config as per usual and enabling it on server start if the config value exists?

lelemm commented:
> > This script would trigger what the UI triggers on the server, but instead of passing each parameter, it would look for the values in the environment or config.json It would not walk the user through, but instead validate and enable based on the environment/config
> 
> What's the downside of loading the config as per usual and enabling it on server start if the config value exists?

Detecting that you are in a state that needs to run it is the downside, specially for disabling it
(I guess would be two commands, enable and disable)

MikesGlitch commented:
> Detecting that you are in a state that needs to run it is the downside, specially for disabling it (I guess would be two commands, enable and disable)

I'll defer to your judgement at this point. If you think this is best go for it and we can review after.

lelemm commented:
Ok. Thanks. I'll implement the two commands when possible.

youngcw commented:
> Ok. Thanks. I'll implement the two commands when possible.

Merge freeze is in a few days and it would be nice to get this is before then if you can get those functions added.

lelemm commented:
> > Ok. Thanks. I'll implement the two commands when possible.
> 
> Merge freeze is in a few days and it would be nice to get this is before then if you can get those functions added.

@youngcw, this two commands will be on server repo. I did the commands already, I just need to test more.
We can keep the review on this PR parallelly

matt-fidd commented:
Great work on this, thank you for all of the amendments so far.

These suggestions are far more subjective than functional ones, but there a few visual nits that would be nice to look at.

## Login select box
![image](https://github.com/user-attachments/assets/083fa382-6841-4145-8393-1399758fc874)

Something feels a little off here, I think that the select should be the same size as the "sign in" button if possible. 

## Transfer ownership menu
![image](https://github.com/user-attachments/assets/d552326a-240d-407c-b78c-4be45bce63ff)

This is quite hidden, and the purpose is vague until you click onto the lock icon. If I was a user looking for this functionality, I'd probably take a long time or have to ask for help.

Some potential alternatives:
- It might be better as a primary button at the bottom of the page
- Use the tick boxes already present at the right hand side of the table (with a confirmation modal)
- Add a three dot menu to each row of the table which then shows options for that user such as "make owner"

## Transfer ownership modal
![image](https://github.com/user-attachments/assets/4fce38f5-c8a0-47c2-abcb-a86869f00383)

The select element should have placeholder text

Text starting with "This action is irreversable" would be better in a new paragraph and highlighted red.
![image](https://github.com/user-attachments/assets/8871a19e-4ea6-4b33-9b9b-c81155f76eee)

The text would read more clearly if it was reworded as:
`This action is irreversible, ownership of this budget file will only be able to be transferred by the server administrator or new owner.`

![image](https://github.com/user-attachments/assets/1ad7187b-677b-43a4-8ecc-82d1df74f91a)
The "transfer ownership" button should be purple

## User edit modal
![image](https://github.com/user-attachments/assets/23864c30-09df-4002-b6a3-81c8249f5e38)

Instead of "user" as the title, is it possible to use the username of the user that's being edited?

## Revert to server password modal
![image](https://github.com/user-attachments/assets/b5eba22f-d527-4f4c-ad0b-176f92c81230)

The password input box would look better with a border, it blends in to the background at the moment

It should be made more clear that you need to enter the current server password to confirm the change, at the moment it just prompts for "a password" which could be misinterpreted as setting a new one.

If you enter an incorrect password, it shows "internal server error"
![image](https://github.com/user-attachments/assets/a704d191-3e03-4e46-a4bc-8405a242391c)


## General nits
There are a few buttons missing their borders, especially in the modals. Below is an example from the "close account" modal of what the buttons should look like.
![image](https://github.com/user-attachments/assets/f1c75ea0-24ef-46f5-adfa-2393a1ad3848)

Using the warning text colour for the caution messages for destruction actions makes sense, but maybe red would be more fitting to better translate "danger"
![image](https://github.com/user-attachments/assets/00f8a9c5-212f-47dc-8611-c419dc09566b)
vs
![image](https://github.com/user-attachments/assets/3cb87c50-2255-4347-8a7d-93990c04f30d)

lelemm commented:
@matt-fidd thanks for the feedback. I will try to address ao this asap.

this error:
![image](https://github.com/user-attachments/assets/dcf514f8-89bb-461d-942c-e8317fbb4611)

its already solved on the latest server.

matt-fidd commented:
> @matt-fidd thanks for the feedback. I will try to address all this asap.
> 
> 
> 
> this error:
> 
> ![image](https://github.com/user-attachments/assets/dcf514f8-89bb-461d-942c-e8317fbb4611)
> 
> 
> 
> its already solved on the latest server.

Ah sorry, I pulled the new client but not server

lelemm commented:
@matt-fidd changed all the items. please, check it now

matt-fidd reviewed (approved):
Thanks for this! Here's my +1 for getting it in this release

youngcw commented:
@MikesGlitch any more feedback from you?  Ill merge this later if I don't see any new comments.

MikesGlitch reviewed (approved):
Thanks for all the great work  @lelemm  🎉

lelemm commented:
Nice. We need to work are the docs. There will be many questions about this feature need release,

lgwapnitsky commented:
excited for this to make it to the latest docker pull :)
